pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+from __future__ import annotations
+
+import contextlib
+import hashlib
+import os.path
+import platform
+import stat
+from base64 import standard_b64decode as b64decode
+from base64 import standard_b64encode as b64encode
+from dataclasses import dataclass
+from os import chmod
+from pathlib import Path
+from pathlib import PurePath
+from types import MappingProxyType
+from typing import Final
+from typing import Generator
+from typing import Sequence
+from urllib.parse import urlparse
+from urllib.request import urlopen
+
+from pre_commit import lang_base
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.prefix import Prefix
+
+ENVIRONMENT_DIR = 'download'
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+IO_BUF_SIZE = 4096
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+class Platform:
+    OS: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'Linux': 'linux',
+        'Darwin': 'darwin',
+        'Windows': 'windows',
+        'DragonFly': 'dragonfly',
+        'FreeBSD': 'freebsd',
+    })
+    CPU: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'aarch64': 'arm64',
+        'aarch64_be': 'arm64be',
+        'arm': 'arm',
+        'i386': '386',
+        'i686': '386',
+        'x86_64': 'amd64',
+        'AMD64': 'amd64',
+        'ppc': 'ppc',
+        'ppc64': 'ppc64',
+        'ppc64le': 'ppc64le',
+    })
+
+    def __init__(self, value: str) -> None:
+        self.__value = value
+        os, cpu = self.parts
+        if os not in Platform.OS.values():
+            raise ValueError(f'invalid operating system `{os}`')
+        if cpu not in Platform.CPU.values():
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> str:
+        os, _ = self.parts
+        return os
+
+    @property
+    def cpu(self) -> str:
+        _, cpu = self.parts
+        return cpu
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @staticmethod
+    def host() -> Platform:
+        os = Platform.OS[platform.system()]
+        cpu = Platform.CPU[platform.machine()]
+        return Platform(f'{os}/{cpu}')
```

Comment:
It is good DRY practice to use `@classmethod` here:

```suggestion
    @classmethod
    def host(cls: Platform) -> Platform:
        os = cls.OS[platform.system()]
        cpu = cls.CPU[platform.machine()]
        return cls(f'{os}/{cpu}')
```

The `Platform` can be removed with [Self](https://docs.python.org/3/library/typing.html#typing.Self) but that is not available until 3.11. It is in `typing_extensions` but `pre-commit` does not use that dependency. So this snippet could be:

```py
    @classmethod
    def host(cls: Self) -> Self:
        os = cls.OS[platform.system()]
        cpu = cls.CPU[platform.machine()]
        return cls(f'{os}/{cpu}')
```

Reply:
If I do that, I get these:
```
pre_commit/languages/download.py:99: error: The erased type of self "pre_commit.languages.download.Platform" is not a supertype of its class "Type[pre_commit.languages.download.Platform]"  [misc]
pre_commit/languages/download.py:102: error: "Platform" not callable  [operator]
pre_commit/languages/download.py:250: error: Invalid self argument "type[Platform]" to attribute function "host" with type "Callable[[Platform], Platform]"  [misc]
Found 3 errors in 1 file (checked 134 source files)
```
The line numbers refer to this particular change. I am not sure how to interpret these. I am happy to have a call to get a better understanding of why this is happening.

Reply:
That error is correct, my suggestion is wrong.

`@classmethod` thats the _type_ of the class not the instance.

So it should be `typing.Type[Platform]`:

```suggestion
    @classmethod
    def host(cls: Type[Platform]) -> Platform:
        os = cls.OS[platform.system()]
        cpu = cls.CPU[platform.machine()]
        return cls(f'{os}/{cpu}')
```

Reply:
Done